_posts/2020-09-31-spring-component-scanning.md

```diff
@@ -0,0 +1,252 @@
+
+### 2. `@Controller`
+
+This indicates that the annotated class is a Spring-managed controller which is used with annotated handler methods based on `@RequestMapping`. And Spring 4.0 introduced `@RestController` which combines both `@Controller` and `@ResponseBody` and makes it easy to create RESTful services.
+
+### 3. `@Service`
+
+We can use the `@Service` stereotype for classes that contains the business logic or the classes which come in the service layer.
+
+### 4. `@Respository`
+
+We can use the `@Repository` stereotype for DAO classes which are responsible for providing CRUD operations on entities. If we are using Spring Data for managing database operations, then we should use the Spring Data Repository interface.
+
+## When to Use Component Scanning
+
+Spring provides a mechanism to identify managed components explicitly through `@ComponentScan` annotation, we'll look into various properties which we can use in further sections.
+
+For a normal Spring application, we can tell Spring to perform a component scan by explicitly specifying in XML configuration or Java Configuration.
+
+If the application is a Spring Boot application, then all the packages under the parent package will be covered by an implicit component scan. Spring Boot's `@SpringBootApplication` annotation is comprised of `@Configuration`, `@ComponentScan`, and `@EnableAutoConfiguration` annotations. As we mentioned it includes `@ComponentScan`, which will scans for the components in the current package and all its sub-packages. So if your application doesn't have a varying package structure then there is no need for explicit component scanning.
+
+**Specifying `@Configuration` in the default package will tell Spring to scan all the classes in all the jars in the classpath.**
+
+## How to Use `@ComponentScan`
+
+`@ComponentScan` along with `@Configuration` annotation is used to tell Spring to scan classes that are annotated with stereotype annotation. The `@ComponentScan` annotation provides different attributes that we can modify to get desired scanning behavior.
+
+```java
+@Component
+public class BeanViewer {
+
+  private final Logger LOG = LoggerFactory.getLogger(getClass());
+
+  @EventListener
+  public void showBeansRegistered(ApplicationReadyEvent event) {
+    String[] beanNames = event.getApplicationContext().getBeanDefinitionNames();
+      for(String beanName: beanNames) {
+        LOG.info("{}", beanName);
+    }
+  }
+}
+```
+
+We'll be using `ApplicationContext`'s `getBeanDefinitionNames()` method throughout this article to check out the list of beans created under the application context.
+
+The above `BeanViewer` will print all the beans that are registered with the application context. This will help us to check whether our components are loaded properly or not.
+
+### Spring Boot's Implicit Auto Scanning
```

Comment:
This section shows that Spring Boot's auto scanning works, but not why it works. Please add a code example that shows the code of the `@SpringBootApplication` annotation and explain how it is meta-annotated with `@ComponentScan`.

Reply:
Actually I have explained that in "When to Use Component Scanning Section"
--------------------
If the application is a Spring Boot application, then all the packages under the package containing the Spring Boot application class will be covered by an implicit component scan. 

Spring Boot's `@SpringBootApplication` annotation implies the `@Configuration`, `@ComponentScan`, and `@EnableAutoConfiguration` annotations.